homeassistant/components/mobile_app/sensor.py

```diff
@@ -84,23 +83,9 @@ async def async_restore_last_state(self, last_state: State) -> None:
         """Restore previous state."""
         await super().async_restore_last_state(last_state)
         config = self._config
-        if not (last_sensor_data := await self.async_get_last_sensor_data()):
-            # Workaround to handle migration to RestoreSensor, can be removed
-            # in HA Core 2023.4
-            config[ATTR_SENSOR_STATE] = None
-            webhook_id = self._entry.data[CONF_WEBHOOK_ID]
-            if TYPE_CHECKING:
-                assert self.unique_id is not None
-            sensor_unique_id = _extract_sensor_unique_id(webhook_id, self.unique_id)
-            if (
-                self.device_class == SensorDeviceClass.TEMPERATURE
-                and sensor_unique_id == "battery_temperature"
-            ):
-                config[ATTR_SENSOR_UOM] = UnitOfTemperature.CELSIUS
-        else:
+        if last_sensor_data := await self.async_get_last_sensor_data():
             config[ATTR_SENSOR_STATE] = last_sensor_data.native_value
             config[ATTR_SENSOR_UOM] = last_sensor_data.native_unit_of_measurement
```

Comment:
We should add coverage to make sure these get restored as it clear it was missing since removing it didn't cause the test to fail

Reply:
Aren't they covered here?
https://github.com/home-assistant/core/blob/6e634a969bb931c5026de40b331539fdb1e15946/tests/components/mobile_app/test_sensor.py#L124-L125

Reply:
It appears the test is ineffective as it still passes with this change

```diff
diff --git a/homeassistant/components/mobile_app/sensor.py b/homeassistant/components/mobile_app/sensor.py
index 29a310c1c19..e9dcef0edb4 100644
--- a/homeassistant/components/mobile_app/sensor.py
+++ b/homeassistant/components/mobile_app/sensor.py
@@ -82,10 +82,6 @@ class MobileAppSensor(MobileAppEntity, RestoreSensor):
     async def async_restore_last_state(self, last_state: State) -> None:
         """Restore previous state."""
         await super().async_restore_last_state(last_state)
-        config = self._config
-        if last_sensor_data := await self.async_get_last_sensor_data():
-            config[ATTR_SENSOR_STATE] = last_sensor_data.native_value
-            config[ATTR_SENSOR_UOM] = last_sensor_data.native_unit_of_measurement
         self._async_update_attr_from_config()
 
     def _calculate_native_value(self) -> StateType | date | datetime:
```

Reply:
That's why I orignally removed it. I'm confused now as I thought it was restoring from here

https://github.com/home-assistant/core/blob/55134e23ea780cc908998a0602af04b065074bd8/homeassistant/components/mobile_app/entity.py#L72-L76

Reply:
That's not confusing at all right?

It looks like it uses the base restore code there but the sensor specific code in sensor so whats missing is a test for the UOM being restored which is why we can remove the code and it doesn't fail.

Reply:
discussed on discord. summary, we need to handle the case where the UOM is not the native UOM because its been overridden in the registry and the config wants the native UOM before its converted via https://github.com/home-assistant/core/blob/f353b3fa5410842e56b69b096c204d102e6c218f/homeassistant/components/sensor/__init__.py#L508